### PR TITLE
New Caveats #6255

### PR DESCRIPTION
**Pull request by rcipkins**
Added new caveats for new releases and expiration time of a macaroon.

### CODE REVIEW

**Comment by woodruffw** [comment]
Whoops, requested the wrong reviewer -- that should be me.

**Comment by woodruffw** [comment]
This is looking really good! I'll do another pass tomorrow.

File: dev/notdatadog.py [hunk 1]
```diff
@@ -16,11 +16,10 @@
```
Comment (woodruffw) [inline dev/notdatadog.py]: I think it'd be safe to revert this file.

File: dev/smtp.py [hunk 2]
```diff
@@ -16,17 +16,15 @@
```
Comment (woodruffw) [inline dev/smtp.py]: Ditto here.

File: docs/conf.py [hunk 3]
```diff
@@ -28,11 +28,7 @@
 # Add any Sphinx extension module names here, as strings. They can be
```
Comment (woodruffw) [inline docs/conf.py]: Ditto here.

File: tests/unit/manage/test_views.py [hunk 4]
```diff
@@ -1465,9 +1465,11 @@ def test_default_response(self, monkeypatch):
         monkeypatch.setattr(views, "DeleteMacaroonForm", delete_macaroon_cls)
 
         project_names = [pretend.stub()]
```
Comment (woodruffw) [inline tests/unit/manage/test_views.py]: I think this variable + associated monkeypatch (and all duplicates below) can be removed.

File: warehouse/macaroons/caveats.py [hunk 5]
```diff
@@ -42,33 +48,75 @@ def verify_projects(self, projects):
             )
 
         project = self.verifier.context
-        if project.normalized_name in projects:
-            return True
 
+        for user_proj in projects:
```
Comment (woodruffw) [inline warehouse/macaroons/caveats.py]: I think we need to make this loop more flexible to handle API tokens that are already out there in the world: we should check if each element is a `str` -- if it is, then treat is as the normalized name. If it's a `dict`, then perform the `project-name` lookup.

File: warehouse/manage/forms.py [hunk 6]
```diff
@@ -238,12 +251,59 @@ def validate_token_scope(self, field):
...
+        expiration = datetime.strftime(expiration, "%Y-%m-%dT%H:%M")
+
+    def validate(self):
+        res = super().validate()
+        if not isinstance(self.validated_scope, str):
+            self.validated_scope = {
+                "expiration": self.expiration.data,
+                "projects": [
+                    {
+                        "project-name": self.token_scope.data,
```
Comment (woodruffw) [inline warehouse/manage/forms.py]: Could be misunderstanding here, but won't this embed the full `scope:` prefix into the `project-name`? I think we need to parse it out first.

**Comment by woodruffw** [comment]
@ewdurbin @dstufft I believe this is good for an initial review!

Some notes:

* We should probably (better) figure out what the semantics of a release-limited token are -- the upload endpoint accepts one file at a time and creates the release on the first file, so creating a version-limited token will cause package upload to bomb out if the release is more than one file.

* Similarly, PyPI already imposes some release constraints: duplicate filenames can never be re-released. This partially obviates the general need for version-scoped tokens, although it doesn't cover the case where a malicious user uploads a new filename for the same release.

**Comment by ewdurbin** [comment]
ref #6255

**Comment by woodruffw** [comment]
#6935 supersedes this and includes additional caveat work; thanks @rcipkins!

**Pull request closed without merging**